### PR TITLE
Regression test on infinite loop with synchronous useEffect + setState in render 

### PR DESCRIPTION
**Pull request by tyao1**
## Summary

This is a regression test to demonstrate a situation that React would lock up the JS thread in the infinite loop. 
it doesn't break the infinite loop because the Maximum update exception is caught and handled in `recoverFromConcurrentError` of `performSyncWorkOnRoot`, and succeeds in the retry, which doesn't break the loop and makes the exception a recoverable error.

- [ ]  TODO: figure out if/how to fix it

### CODE REVIEW

File: packages/react-dom/src/__tests__/ReactUpdates-test.js [hunk 1]
```diff
@@ -1734,4 +1739,57 @@ describe('ReactUpdates', () => {
...
+    
+    function B({stateRef}) {
+      const {updateA, a} = stateRef.current;
+      const [b, updateB] = React.useState(false);
+      stateRef.current.updateB = updateB;
+      React.useEffect(() => {
+        updateB(false);
+      });
+      if (updateA) {
+        updateA(!a);
```
Comment (acdlite) [inline packages/react-dom/src/__tests__/ReactUpdates-test.js:1768]: This is accompanied by a warning, right?

Comment (tyao1) [reply packages/react-dom/src/__tests__/ReactUpdates-test.js:1768]: There is a `Maximum update depth exceeded` exception on this call, it is thrown as a recoverable error by `onRecoverableError`.

File: packages/react-dom/src/__tests__/ReactUpdates-test.js [hunk 2]
```diff
@@ -1734,4 +1739,57 @@ describe('ReactUpdates', () => {
       });
     }).toThrow('Maximum update depth exceeded');
   });
+
+  fit('prevents infinite update loop between a component with synchronous updates in useEffect and a component with synchronous updates in render', async () => {
```
Comment (alunyov) [inline packages/react-dom/src/__tests__/ReactUpdates-test.js:1743]: nit: should be `it`? Is there a lint rule that checks these?

File: packages/react-dom/src/__tests__/ReactUpdates-test.js [hunk 3]
```diff
@@ -25,6 +26,10 @@ describe('ReactUpdates', () => {
     ReactTestUtils = require('react-dom/test-utils');
     act = require('jest-react').act;
     Scheduler = require('scheduler');
+    ReactFeatureFlags = require('shared/ReactFeatureFlags');
+    ReactFeatureFlags.enableSyncDefaultUpdates = true;
+    ReactFeatureFlags.enableUnifiedSyncLane = true;
+    ReactFeatureFlags.enableUseRefAccessWarning = false;
```
Comment (alunyov) [inline packages/react-dom/src/__tests__/ReactUpdates-test.js:32]: do we need these flags for all tests, or they only for specific case(maybe we should move these flags in that test?)

File: packages/react-dom/src/__tests__/ReactUpdates-test.js (hunk 2 above)
Comment (alunyov) [reply packages/react-dom/src/__tests__/ReactUpdates-test.js:1743]: oh... this is a Draft PR. Sorry for these early comments, I'm still adjusting to this workflow.

File: packages/react-dom/src/__tests__/ReactUpdates-test.js (hunk 3 above)
Comment (tyao1) [reply packages/react-dom/src/__tests__/ReactUpdates-test.js:32]: no, these need to be cleaned up.

**Pull request closed without merging**